ocaml/typing/typecore.ml

```diff
@@ -5058,6 +5058,38 @@ let add_check_attribute expr attributes =
     end
   | _ -> expr
 
+let zero_alloc_of_application ~num_args attrs funct =
+  let zero_alloc =
+    Builtin_attributes.get_property_attribute ~in_signature:false
+      ~default_arity:num_args attrs Zero_alloc
+  in
+  let zero_alloc =
+    match zero_alloc with
+    | Ignore_assert_all _ | Check _ | Assume _ ->
+      (* The user wrote a zero_alloc attribute on the application - keep it *)
```

Comment:
I think that `Ignore_assert_all _`  and `Check _` are not allowed here, and if so we should have  `fatal_error` here.  For user provided `Assume _`, do we have a test to make sure it is not dropped when the signature is annotated?

Reply:
> I think that `Ignore_assert_all _` and `Check _` are not allowed here, and if so we should have `fatal_error` here.

You're right that they aren't allowed, but that is checked in the call to `assume_zero_alloc` below, so putting fatal_error here would result in bad errors.  I have added a comment explaning that.

> For user provided `Assume _`, do we have a test to make sure it is not dropped when the signature is annotated?

Good idea - added.